### DISCUSSION TITLE
Using S3 as a backend for storage

### DISCUSSION THREAD

**Discussion by tutiplain** [codenotary/immudb#1140] [Q&A] [answered]
Hi all,

  I've been testing using S3 as a backend for ImmuDB. But what I'm noting is that even when I set the options for S3 storage, the local storage is still used (right now I'm using the env variables as described in the docs), and the data is only written to S3 when the immudb server is shutdown. Is the S3 storage feature only meant to backup the database, rather than to be used as the main storage medium? Is there a way to get it to write to S3 after , say,  every few minutes, or perhaps after every commit? Any info on this will be greatly appreciated. Thanks in anticipation. Can S3 storage be used with the embedded version?

**Comment by jaboutboul** [comment]
Hello @tutiplain. The S3 storage backend is not meant solely as a backup feature and this should work. Can you please share more information or some logs perhaps? Do you have cloudwatch enabled on the bucket in S3? What you are trying to do should definitely work so I am wondering is its just a configuration issue somewhere.

**Reply by tutiplain** [reply]
Hi, 

  Thanks for replying.  Right now I am using immudb server on my local dev machine, and using MinIO as my S3 backend. The MInIO instance is also running locally on my machine. I have set up the following environment variables on my machine:

`IMMUDB_S3_BUCKET_NAME=immu
IMMUDB_S3_ACCESS_KEY_ID=minioadmin
IMMUDB_S3_LOCATION=
IMMUDB_S3_STORAGE=true
IMMUDB_S3_ENDPOINT=http://localhost:9000
IMMUDB_S3_SECRET_KEY=minioadmin`

I then run immudb server with the following command:
`./immudb-v1.2.2-darwin-amd64 --pgsql-server=false`


The terminal output shows the following:

` 
immudb 1.2.2
Commit  : 89c3e3e8a4293909421d23025d7a26bb04b35a5e
Built by: cleaversdev@gmail.com
Built at: Tue, 18 Jan 2022 10:27:58 AST
================ Config ================
Data dir         : ./data
Address          : 0.0.0.0:3322
Metrics address  : 0.0.0.0:9497/metrics
Config file      : configs/immudb.toml
Max recv msg size: 33554432
Auth enabled     : true
Dev mode         : false
Default database : defaultdb
Maintenance mode : false
Synced mode      : true
S3 storage
   endpoint      : http://localhost:9000
   bucket name   : immu
   prefix        : 
----------------------------------------
Superadmin default credentials
   Username      : immudb
   Password      : immudb
========================================

2022/02/25 12:27:18 S3 HEAD http://localhost:9000/immu/immudb.identifier
2022/02/25 12:27:18 S3 HEAD http://localhost:9000/immu/immudb.identifier 404 Not Found
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix= 200 OK
immudb  2022/02/25 12:27:18 INFO: Opening database 'systemdb' {replica = false}...
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/systemdb/tx/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Ftx%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Ftx%2F 200 OK
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/systemdb/commit/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Fcommit%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Fcommit%2F 200 OK
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/systemdb/val_0/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Fval_0%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Fval_0%2F 200 OK
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/systemdb/aht/data/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Faht%2Fdata%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Faht%2Fdata%2F 200 OK
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/systemdb/aht/tree/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Faht%2Ftree%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Faht%2Ftree%2F 200 OK
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/systemdb/aht/commit/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Faht%2Fcommit%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Faht%2Fcommit%2F 200 OK
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/systemdb/index/history/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Findex%2Fhistory%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Findex%2Fhistory%2F 200 OK
immudb  2022/02/25 12:27:18 INFO: Reading snapshot at 'data/systemdb/index/commit'...
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/systemdb/index/nodes/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Findex%2Fnodes%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Findex%2Fnodes%2F 200 OK
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/systemdb/index/commit/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Findex%2Fcommit%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=systemdb%2Findex%2Fcommit%2F 200 OK
immudb  2022/02/25 12:27:18 INFO: Successfully read snapshot at 'data/systemdb/index/commit'
immudb  2022/02/25 12:27:18 INFO: Indexing in progress at 'data/systemdb'
immudb  2022/02/25 12:27:18 INFO: Binary Linking up to date at 'data/systemdb'
immudb  2022/02/25 12:27:18 INFO: Database 'systemdb' {replica = false} successfully opened
immudb  2022/02/25 12:27:18 INFO: Loading SQL Engine for database 'systemdb' {replica = false}...
immudb  2022/02/25 12:27:18 INFO: Flushing index 'data/systemdb/index'...
immudb  2022/02/25 12:27:18 INFO: Opening database 'defaultdb' {replica = false}...
immudb  2022/02/25 12:27:18 INFO: Flushing not needed at 'data/systemdb/index'
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/defaultdb/tx/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Ftx%2F
immudb  2022/02/25 12:27:18 INFO: SQL Engine ready for database 'systemdb' {replica = false}
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Ftx%2F 200 OK
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/defaultdb/commit/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Fcommit%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Fcommit%2F 200 OK
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/defaultdb/val_0/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Fval_0%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Fval_0%2F 200 OK
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/defaultdb/aht/data/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Faht%2Fdata%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Faht%2Fdata%2F 200 OK
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/defaultdb/aht/tree/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Faht%2Ftree%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Faht%2Ftree%2F 200 OK
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/defaultdb/aht/commit/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Faht%2Fcommit%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Faht%2Fcommit%2F 200 OK
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/defaultdb/index/history/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Findex%2Fhistory%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Findex%2Fhistory%2F 200 OK
immudb  2022/02/25 12:27:18 INFO: Reading snapshot at 'data/defaultdb/index/commit'...
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/defaultdb/index/nodes/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Findex%2Fnodes%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Findex%2Fnodes%2F 200 OK
2022/02/25 12:27:18 Opening remote storage at s3:http://localhost:9000/immu/defaultdb/index/commit/
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Findex%2Fcommit%2F
2022/02/25 12:27:18 S3 GET http://localhost:9000/immu/?delimiter=%2F&encoding-type=url&list-type=2&prefix=defaultdb%2Findex%2Fcommit%2F 200 OK
immudb  2022/02/25 12:27:18 INFO: Successfully read snapshot at 'data/defaultdb/index/commit'
immudb  2022/02/25 12:27:18 INFO: Indexing in progress at 'data/defaultdb'
immudb  2022/02/25 12:27:18 INFO: Binary Linking up to date at 'data/defaultdb'
immudb  2022/02/25 12:27:18 INFO: Database 'defaultdb' {replica = false} successfully opened
immudb  2022/02/25 12:27:18 INFO: Loading SQL Engine for database 'defaultdb' {replica = false}...
immudb  2022/02/25 12:27:18 INFO: Flushing index 'data/defaultdb/index'...
immudb  2022/02/25 12:27:18 INFO: Flushing not needed at 'data/defaultdb/index'
2022/02/25 12:27:18 S3 PUT http://localhost:9000/immu/immudb.identifier
immudb  2022/02/25 12:27:18 INFO: SQL Engine ready for database 'defaultdb' {replica = false}
2022/02/25 12:27:18 S3 PUT http://localhost:9000/immu/immudb.identifier 200 OK
immudb  2022/02/25 12:27:19 INFO: Webconsole enabled: 0.0.0.0:8080
immudb  2022/02/25 12:27:19 INFO: Web API server enabled on 0.0.0.0:8080/api (http)
You can now use immuadmin and immuclient CLIs to login with the immudb superadmin user and start using immudb.`

At this time, I can see in my bucket that a file named immudb.identifier is created. I also see that a "data" folder is created on my machine, in the same folder where the immudb executable is. I then proceed to execute some queries to create a table and insert some records. I can query the data on the web console, but my bucket remains unchanged. So I then proceed to shutdown immudb by pressing Control-C.  I see the following on the terminal:

`immudb  2022/02/25 12:32:24 INFO: 1 transaction/s to be indexed at 'data/defaultdb'
immudb  2022/02/25 12:32:24 INFO: Flushing index 'data/defaultdb/index'...
immudb  2022/02/25 12:32:24 INFO: Flushing index 'data/defaultdb/index' successfully completed
^Cimmudb  2022/02/25 12:34:22 INFO: Caught SIGTERM
immudb  2022/02/25 12:34:22 INFO: Stopping immudb:
================ Config ================
Data dir         : ./data
Address          : 0.0.0.0:3322
Metrics address  : 0.0.0.0:9497/metrics
Config file      : configs/immudb.toml
Max recv msg size: 33554432
Auth enabled     : true
Dev mode         : false
Default database : defaultdb
Maintenance mode : false
Synced mode      : true
S3 storage
   endpoint      : http://localhost:9000
   bucket name   : immu
   prefix        : 
----------------------------------------
Superadmin default credentials
   Username      : immudb
   Password      : immudb
========================================
immudb  2022/02/25 12:34:22 INFO: sessions guard started
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/val_0/00000000.val
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/val_0/00000000.val 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/val_0/00000000.val
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/val_0/00000000.val 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/val_0/00000000.val range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/val_0/00000000.val
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/val_0/00000000.val 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
immudb  2022/02/25 12:34:22 INFO: Indexing gracefully stopped at 'data/defaultdb'
immudb  2022/02/25 12:34:22 INFO: Flushing index 'data/defaultdb/index'...
immudb  2022/02/25 12:34:22 INFO: Flushing not needed at 'data/defaultdb/index'
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/index/nodes/00000000.n
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/index/nodes/00000000.n 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/index/nodes/00000000.n
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/index/nodes/00000000.n 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/index/nodes/00000000.n range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/index/nodes/00000000.n
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/index/nodes/00000000.n 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/index/history/00000000.hx
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/index/history/00000000.hx 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/index/history/00000000.hx
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/index/history/00000000.hx 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/index/history/00000000.hx range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/index/history/00000000.hx
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/index/history/00000000.hx 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/index/commit/00000000.ri
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/index/commit/00000000.ri 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/index/commit/00000000.ri
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/index/commit/00000000.ri 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/index/commit/00000000.ri range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/index/commit/00000000.ri
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/index/commit/00000000.ri 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/tx/00000000.tx
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/tx/00000000.tx 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/tx/00000000.tx
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/tx/00000000.tx 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/tx/00000000.tx range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/tx/00000000.tx
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/tx/00000000.tx 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/commit/00000000.txi
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/commit/00000000.txi 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/commit/00000000.txi
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/commit/00000000.txi 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/commit/00000000.txi range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/commit/00000000.txi
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/commit/00000000.txi 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/aht/data/00000000.dat
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/aht/data/00000000.dat 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/aht/data/00000000.dat
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/aht/data/00000000.dat 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/aht/data/00000000.dat range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/aht/data/00000000.dat
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/aht/data/00000000.dat 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/aht/tree/00000000.sha
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/aht/tree/00000000.sha 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/aht/tree/00000000.sha
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/aht/tree/00000000.sha 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/aht/tree/00000000.sha range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/aht/tree/00000000.sha
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/aht/tree/00000000.sha 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/aht/commit/00000000.di
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/defaultdb/aht/commit/00000000.di 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/aht/commit/00000000.di
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/defaultdb/aht/commit/00000000.di 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/aht/commit/00000000.di range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/aht/commit/00000000.di
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/defaultdb/aht/commit/00000000.di 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/val_0/00000000.val
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/val_0/00000000.val 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/val_0/00000000.val
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/val_0/00000000.val 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/val_0/00000000.val range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/val_0/00000000.val
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/val_0/00000000.val 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
immudb  2022/02/25 12:34:22 INFO: Indexing gracefully stopped at 'data/systemdb'
immudb  2022/02/25 12:34:22 INFO: Flushing index 'data/systemdb/index'...
immudb  2022/02/25 12:34:22 INFO: Flushing not needed at 'data/systemdb/index'
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/index/nodes/00000000.n
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/index/nodes/00000000.n 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/index/nodes/00000000.n
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/index/nodes/00000000.n 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/index/nodes/00000000.n range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/index/nodes/00000000.n
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/index/nodes/00000000.n 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/index/history/00000000.hx
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/index/history/00000000.hx 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/index/history/00000000.hx
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/index/history/00000000.hx 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/index/history/00000000.hx range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/index/history/00000000.hx
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/index/history/00000000.hx 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/index/commit/00000000.ri
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/index/commit/00000000.ri 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/index/commit/00000000.ri
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/index/commit/00000000.ri 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/index/commit/00000000.ri range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/index/commit/00000000.ri
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/index/commit/00000000.ri 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/tx/00000000.tx
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/tx/00000000.tx 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/tx/00000000.tx
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/tx/00000000.tx 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/tx/00000000.tx range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/tx/00000000.tx
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/tx/00000000.tx 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/commit/00000000.txi
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/commit/00000000.txi 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/commit/00000000.txi
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/commit/00000000.txi 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/commit/00000000.txi range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/commit/00000000.txi
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/commit/00000000.txi 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/aht/data/00000000.dat
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/aht/data/00000000.dat 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/aht/data/00000000.dat
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/aht/data/00000000.dat 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/aht/data/00000000.dat range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/aht/data/00000000.dat
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/aht/data/00000000.dat 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/aht/tree/00000000.sha
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/aht/tree/00000000.sha 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/aht/tree/00000000.sha
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/aht/tree/00000000.sha 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/aht/tree/00000000.sha range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/aht/tree/00000000.sha
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/aht/tree/00000000.sha 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/aht/commit/00000000.di
2022/02/25 12:34:22 S3 PUT http://localhost:9000/immu/systemdb/aht/commit/00000000.di 200 OK
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/aht/commit/00000000.di
2022/02/25 12:34:22 S3 HEAD http://localhost:9000/immu/systemdb/aht/commit/00000000.di 200 OK
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/aht/commit/00000000.di range: 0 -1
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/aht/commit/00000000.di
2022/02/25 12:34:22 S3 GET http://localhost:9000/immu/systemdb/aht/commit/00000000.di 206 Partial Content
2022/02/25 12:34:22 Uploadnig of chunk 0 finished in state: Remote
immudb  2022/02/25 12:34:22 INFO: Shutdown completed`

But now, the S3 bucket has the same content as the "data" folder in my local machine. Because I see no change in the bucket until after shutting down immudb, that is why I asked if S3 is only meant as a backup.  Is there any additional configuration that I need to do to get S3 sync without shutting down immudb?

**Comment by jeroiraz** [comment, marked as answer]
Hi @tutiplain, immudb uses a multi-file approach. Files are created with exactly the same size as they are filled up. Once a file is fully written and a new file is created for the next writes, the previous one will be automatically uploaded to the S3 server. This upload also takes place when gracefully closing immudb server

**Comment by vchaindz** [comment]
Hi @tutiplain if you want to force immudb to write all local data to S3, you need to gracefully shutdown immudb. As @jeroiraz mentioned, files are automatically synced once they reach a certain size and are filled up (currently 1MB) or when immudb is shutting down.
